tracer/src/Datadog.Trace/Util/Delegates/DelegateInstrumentation.cs

```diff
@@ -0,0 +1,1122 @@
+﻿// <copyright file="DelegateInstrumentation.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+using Datadog.Trace.ClrProfiler.CallTarget.Handlers.Continuations;
+
+namespace Datadog.Trace.Util.Delegates;
+
+#pragma warning disable SA1124
+
+internal class DelegateInstrumentation
+{
+    public static Delegate Wrap<TCallbacks>(Delegate target, TCallbacks callbacks)
+        where TCallbacks : struct, ICallbacks
+        => Wrap<Delegate, TCallbacks>(target, callbacks);
+
+    public static TDelegate Wrap<TDelegate, TCallbacks>(TDelegate target, TCallbacks callbacks)
+        where TDelegate : Delegate
+        where TCallbacks : struct, ICallbacks
+    {
+        var targetType = target.GetType();
+        var invokeMethod = targetType.GetMethod("Invoke");
+        var returnType = invokeMethod!.ReturnType;
+        var arguments = invokeMethod!.GetParameters();
+        switch (arguments.Length)
+        {
+            case 0:
+            {
+                if (returnType == typeof(void))
+                {
+                    var wrapperType = typeof(Action0Wrapper<,>).MakeGenericType(
+                        targetType,
+                        typeof(TCallbacks));
```

Comment:
I'm not really sure how it would look currently, but for instrumentations, and given we _likely_ know all of this up-front, I wonder if we should (and/or) _can_ create something like `ActivatorHelper`, so we only do the majority of this work once, + use a custom activator for best of both worlds 🤔 

Not suggesting we do it in this PR of course! Just that this will (hopefully) mostly be necessary when we're dealing with known types. Also makes me wonder if we can build this directly into the duck typing at some point, but that may be a step too far 😅

Reply:
That's a good idea for another PR 😄

Reply:
done in [03b76a6](https://github.com/DataDog/dd-trace-dotnet/pull/4613/commits/03b76a6fc7e03d47d50fccc5918babce56414bf4)

Reply:
nice 👍 You couldn't resist 😂